### PR TITLE
Move `DependencyGraph` from `//language-support/scala/codegen` to `//language-support/codegen-common/`

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
changelog_begin
changelog_end

The idea is that of re-using `DependencyGraph` for the Java codegen.

Builds on top of https://github.com/digital-asset/daml/pull/13361 and
contributes towards https://github.com/digital-asset/daml/issues/13324.

In order for the extraction to work, I also had to factor out a couple of
functions out of `LFUtil` and put it in a shared `Util` object that is visible
only within the `codegen` package.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+package com.daml.codegen.lf
+
+import com.daml.lf.iface
+
+case class DefTemplateWithRecord[+Type](
+    `type`: iface.Record[Type],
+    template: iface.DefTemplate[Type],
+)
+object DefTemplateWithRecord {
+  type FWT = DefTemplateWithRecord[iface.Type]
```
Comment (ray-roestenburg-da) [inline language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala:13]: What does FWT stand for?  Is it used?

Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala:13]: > What does FWT stand for?

Field With Type? Probably @S11001001 can give more context, I just copied this code.

> Is it used?

Yes, six instances.

---

I would probably have a look and see if this needs to be generic.

**Comment by ray-roestenburg-da** [comment]
Is it codegen common, or codegen-dependency-graph? I like specific lib names if possible?

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala (hunk 1 above)
Comment (S11001001) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala:13]: Stands for `Field With Type`; the design was changed so only the `Type` remains in this usage but the many references to the alias were left. I don't know if this particular alias is used but the alias in general is used a lot more than 6 times.

Comment (S11001001) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala:13]: I would just extend `FWTLike` here by the way. That will make renaming and find-usages a lot more straightforward should we do that in the future.

**Comment by stefanobaghino-da** [comment]
> Is it codegen common, or codegen-dependency-graph? I like specific lib names if possible? (would it make sense to extract this to a specific lib that does something re: dependency graph, instead of adding it to some 'common lib') I never like 'common libs', but I don't know the code base well enough yet to see if this is 'one of those' @stefanobaghino-da

I'm already building on top of this, would you mind if we move past this and I can split the libraries at a later time?

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala:13]: I'm already working on top of this, I can remove this unnecessary generality but I have changes stacked up on top of this and I would need to refactor this at a later time, is that a problem?

**Comment by ray-roestenburg-da** [comment]
Oh sure, yes this was not something I wanted to block

**Review by ray-roestenburg-da — APPROVED** [review]

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/codegen/lf/DefTemplateWithRecord.scala:13]: Addressed by https://github.com/digital-asset/daml/pull/13372 (which builds on top of https://github.com/digital-asset/daml/pull/13370/)

**Pull request closed without merging**